src/ListT/Weighted.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | A re-implementation of @weighted-search@
+-- as a monad transformer.
+module ListT.Weighted
+  ( Weight (..),
+    WeightedT,
+    Weighted,
+    weight,
+    (<|>),
+    toList,
+  )
+where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus (..), ap)
+import Control.Monad.IO.Class (MonadIO (..))
+import Control.Monad.Reader.Class (MonadReader (..))
+import Control.Monad.State.Class (MonadState (..))
+import Control.Monad.Trans (MonadTrans (..))
+import Control.Monad.WeightedSearch (Weight (..))
+import Data.Foldable (fold, toList)
+import Data.Functor.Identity (Identity)
+
+-- | Weighted nondeterminstic computations over the weight @w@.
+data WeightedT w m a
+  = Fail
+  | Yield a (WeightedT w m a)
+  | Weight w (WeightedT w m a)
+  | Action (m (WeightedT w m a))
```

Comment:
Could you please add a comment on this datatype. My intuition of what a `Weighted n a` is, expect this type to be isomorphic to `(n, a)`.

Reply:
I don't see `Weighted w m a` to be isomorphic to `(n , a)`; I think it is closer to a list of values, weights or a monadic value that returns lists of the same form.

Reply:
I added a few comments. The idea is that `Yield` represents a `Cons`: you return an `a` and keep returning more stuff. The `Weight` is what allows you to state that one part of the computation has a different weight than other.

Reply:
So I guess that `Weighted w m a` is related to `m [a]` through permutation of the list elements (and blissful ignorance of the monadic part...)

Reply:
Exactly. The key point is this line in the definition of `mplus`:

```haskell
  Weight w m `mplus` Yield x n = Yield x (Weight w m `mplus` n)
```

which says that "if you have something with a weight on one side, but can yield on the other, just do so".

Reply:
I can read the code and understand it. But when there is a function of type `f :: Weighted Int a -> Weighted Int [a]` what I guess when reading this type is that `f` takes an object of type `x` (with a weight `w`) and outputs a list of such objects (once again with a weight). It is not at all the meaning of the `Weighted` monad. And I think that it is the kind of thing which makes code hard to read.

Reply:
I realize I probably badly expressed myself the first time. I did not intended to say that the type `Weighted w a` you wrote looks isomorphic to `(w, a)`. I wanted to state that if I jump in a new code and see that something is of type `Weighted w a`, I would expect it to be isomorphic to `(w, a)`.

Reply:
Is this mostly a naming problem? If I renamed it to `WeightedSearch w a`, would that give more the idea that this is like a list monad but with weights?

Reply:
What you are doing is (isomorphic to) a list of `(w, a)` with the guarantee (at type-level) that this list is sorted on the "weights" (ie the first elements). I think that `WeightedList` or `WeightedSet` would convey the right intuition about the fact that we are dealing with a bag of elements, with the guarantee that consumption would be done by increasing weights.

Reply:
Following the denotation of `Weighted w a` as a set of `a`s, reasing the type of `f :: Weighted Int a -> Weighted Int [a]` is simple: it takes a set of `a`s and returns a set of list of `a`s. The `a`s themselves have no weights associated to them, the name is a trap! All the weights do is fix the order in which they are returned (I picked a set as a denotation here precisely to ignore that)

Reply:
Yes, it is a naming issue mainly.

Reply:
Renamed in the latest commit, not it's `WeightedListT`.